### PR TITLE
Update to class based views

### PR DESCRIPTION
**Pull request by DrEdi**
All views now in class based style.

### CODE REVIEW

File: main/templates/main/post_edit.html [hunk 1]
```diff
@@ -1,9 +1,8 @@
 {% extends 'main/base.html' %}
 
 {% block content %}
-    <h1>New post</h1>
```
Comment (DrEdi) [inline main/templates/main/post_edit.html:4]: I made it multipurpose for all classes that need form for their work.
So, I've deleted this string and change text in string a bit lower.
I think it's better to have one multipurpose than some only for one need

File: main/views.py [hunk 2]
```diff
@@ -1,102 +1,115 @@
...
+
+    model = Post
+    fields = ['title', 'text']
+    success_url = reverse_lazy('post_list')
+    template_name = 'main/post_edit.html'
+
+    def form_valid(self, form):
+        """Add info to form that were not given from POST request."""
+        form.instance.author = self.request.user
+        return super(NewPost, self).form_valid(form)
```
Comment (Miseracle) [inline main/views.py]: Mb you can use "super.form_valid(form)" syntax?

Comment (DrEdi) [reply main/views.py]: Nope. `type object 'super' has no attribute 'form_valid'`

Comment (webknjaz) [reply main/views.py]: `super()`

File: blog/settings.py [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 import environ
 
-env = environ.Env(DEBUG=(bool, False),)
+env = environ.Env(DEBUG=(bool, True),)
```
Comment (webknjaz) [inline blog/settings.py]: Please revert this change. defaults are better to be production-ready. add `DEBUG=on` to `.env` instead.

File: main/templates/main/post_edit.html (hunk 1 above)
Comment (webknjaz) [reply main/templates/main/post_edit.html:4]: Please read about YAGNI principle.

File: blog/settings.py (hunk 3 above)
Comment (DrEdi) [reply blog/settings.py]: Could I do this when merging? In master branch it's ok. Just for minimize count of commits

Comment (webknjaz) [reply blog/settings.py]: No, you have to react to comments instantly, otherwise you risk forgetting about them.

File: main/templates/main/index.html [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {% extends 'main/base.html'%}
 
 {% block content %}
-    {% for post in posts %}
+    {% for post in object_list %}
```
Comment (webknjaz) [inline main/templates/main/index.html]: It is possible to rename the queryset result objects name for template context in class-based views. Do this.

File: main/models.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def approved_comments(self):
 class Comment(models.Model):
     """Model for comments."""
 
-    post = models.ForeignKey('main.Post', related_name='comments')
+    post = models.ForeignKey('main.Post', related_name='comments', on_delete=models.CASCADE)
```
Comment (webknjaz) [inline main/models.py:36]: How does this change relate to the goal of the PR? (changing function-views with class-based ones)

File: main/templates/main/post_draft_list.html [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {% extends 'main/base.html' %}
 
 {% block content %}
-    {% for post in posts %}
+    {% for post in object_list %}
```
Comment (webknjaz) [inline main/templates/main/post_draft_list.html]: the above comment applies here as well

File: main/views.py [hunk 7]
```diff
@@ -1,102 +1,115 @@
 """All views are here."""
-from django.shortcuts import render, get_object_or_404, redirect
-from .models import Post, Comment
-from .forms import PostForm, CommentForm
 from django.contrib.auth.decorators import login_required
+from django.utils.decorators import method_decorator
+from django.shortcuts import get_object_or_404, redirect
+from django.core.urlresolvers import reverse_lazy
+from django.views.generic import DetailView, ListView, CreateView, UpdateView, View, DeleteView, TemplateView
+from .forms import CommentForm
```
Comment (webknjaz) [inline main/views.py]: I would separate project-imports from installed packages with an empty string.

File: main/views.py (hunk 2 above)
Comment (webknjaz) [reply main/views.py]: @DrEdi in Python3 `super()` is a shortcut for `super(CurrentClass, self)`

File: main/views.py [hunk 8]
```diff
@@ -1,102 +1,115 @@
...
+from .models import Post, Comment
+
+
+class Protected(View):
+    """Protect views that need to show only for authorised users."""
+
+    @method_decorator(login_required)
+    def dispatch(self, *args, **kwargs):
+        """Rewrite standart method and it to decorator."""
+        return super(Protected, self).dispatch(*args, **kwargs)
```
Comment (webknjaz) [inline main/views.py]: `super`'s arguments may be omitted in Python3 and you should pass `self` to dispatch.

File: main/views.py (hunk 2 above)
Comment (webknjaz) [reply main/views.py]: And I think you missed passing `self`

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: blog/settings.py (hunk 3 above)
Comment (DrEdi) [reply blog/settings.py]: Ok )

File: main/models.py (hunk 5 above)
Comment (DrEdi) [reply main/models.py:36]: Sorry, I thought it was a good idea

File: main/views.py (hunk 2 above)
Comment (DrEdi) [reply main/views.py]: If you mean to .form_valid(form) - it's ok.

File: main/models.py (hunk 5 above)
Comment (webknjaz) [reply main/models.py:36]: It may be a good idea if submitted separately

File: main/views.py [hunk 9]
```diff
@@ -1,102 +1,118 @@
...
 
+    model = Post
+    fields = ['title', 'text']
+    success_url = reverse_lazy('post_list')
+    template_name = 'main/post_edit.html'
+
+    def form_valid(self, form):
+        """Add info to form that were not given from POST request."""
+        form.instance.author = self.request.user
+        return super().form_valid(form)
```
Comment (webknjaz) [inline main/views.py]: I think you missed `self.

File: main/models.py (hunk 5 above)
Comment (webknjaz) [reply main/models.py:36]: or explained why it's necessary here/how it helps to achieve the goal of **current task**

File: main/views.py [hunk 10]
```diff
@@ -1,102 +1,118 @@
...
-    return render(request, 'main/post_edit.html', {'form': form})
 
+    model = Post
+    fields = ['title', 'text']
+    success_url = reverse_lazy('post_list')
+    template_name = 'main/post_edit.html'
+
+    def form_valid(self, form):
+        """Add info to form that were not given from POST request."""
+        form.instance.author = self.request.user
```
Comment (webknjaz) [inline main/views.py:49]: I think it shouldn't go here.

Comment (webknjaz) [reply main/views.py:49]: Try bypassing `request` object into `ModelForm`'s initializer like in https://github.com/open-craft-guild/crm-smsfly/blob/master/SMSFlyCRM/SMSApp/forms/task.py#L44-L46

You may set `form_class` instead of `model` in views :)

File: main/views.py [hunk 11]
```diff
@@ -1,102 +1,118 @@
...
+from .models import Post, Comment
+
+
+class Protected(View):
+    """Protect views that need to show only for authorised users."""
 
+    @method_decorator(login_required)
+    def dispatch(self, *args, **kwargs):
+        """Rewrite standart method and it to decorator."""
+        return super().dispatch(self, *args, **kwargs)
```
Comment (webknjaz) [inline main/views.py]: It looks I was wrong and `self` is being passed automatically in this case

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: main/models.py (hunk 5 above)
Comment (DrEdi) [reply main/models.py:36]: I understood, sorry for my fault

File: main/views.py (hunk 11 above)
Comment (DrEdi) [reply main/views.py]: Ok :+1:

File: vagga.yaml [hunk 12]
```diff
@@ -25,9 +25,9 @@ containers:
         dependencies: true
       - !Install
         - libpq
-      - !Py3Requirements requirements/main.txt
+      - !Py3Requirements requirements/production.txt
     environ:
-      DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/mainDB
+      DATABASE_URL: sqlite://vagga:vagga@127.0.0.1:5433/mainDB
```
Comment (webknjaz) [inline vagga.yaml]: Please don't use sqlite DB.

File: vagga.yaml [hunk 13]
```diff
@@ -25,9 +25,9 @@ containers:
         dependencies: true
       - !Install
         - libpq
-      - !Py3Requirements requirements/main.txt
+      - !Py3Requirements requirements/production.txt
```
Comment (webknjaz) [inline vagga.yaml:28]: plz explain this change

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: vagga.yaml (hunk 13 above)
Comment (DrEdi) [reply vagga.yaml:28]: This container need requirements from production container. We run it with whitnoise.
This was my fault before

File: vagga.yaml (hunk 12 above)
Comment (DrEdi) [reply vagga.yaml]: This container will not run with postgres DB. We move all setting to other container. It's not need here anymore.

Comment (webknjaz) [reply vagga.yaml]: WAT?

Comment (DrEdi) [inline vagga.yaml]: It's not working with postgres but works sqlite :D
I don't know why

Comment (DrEdi) [reply vagga.yaml]: Omg, that was bad idea.
I reviewed code, just forget this...

File: main/views.py [hunk 14]
```diff
@@ -1,104 +1,136 @@
...
+
+from .forms import CommentForm, PostForm
+from .models import Post, Comment
+
 
+class Protected(View):
+    """Protect views that need to show only for authorised users."""
 
-def post_list(request):
+    @method_decorator(login_required)
```
Comment (webknjaz) [inline main/views.py:16]: Is it possible to apply `@login_required` to class?

File: main/views.py [hunk 15]
```diff
@@ -1,104 +1,136 @@
...
+
+class PostList(ListView):
     """Show all your Post objects."""
-    posts = Post.objects.filter(published_date__lte=timezone.now())
-    return render(request, 'main/index.html', {'posts': posts})
 
+    context_object_name = 'posts'
+    template_name = 'main/index.html'
+
+    def get_queryset(self):
```
Comment (webknjaz) [inline main/views.py:28]: AFAIR you could just assign `queryset` property.

File: main/views.py [hunk 16]
```diff
@@ -1,104 +1,136 @@
...
+    fields = ['title', 'text']
+    template_name = 'main/post_edit.html'
+
+    def form_valid(self, form):
+        """Add info to form that were not given from POST request."""
+        form.instance.author = self.request.user
+        return super().form_valid(form)
+
+    def post(self, request, *args, **kwargs):
+        """Rewritten post method for saving the form."""
```
Comment (webknjaz) [inline main/views.py]: This isn't strictly valid description

File: main/views.py [hunk 17]
```diff
@@ -1,104 +1,136 @@
...
             post.save()
             return redirect('post_detail', pk=post.pk)
-    else:
-        form = PostForm()
-    return render(request, 'main/post_edit.html', {'form': form})
 
 
-@login_required
-def post_edit(request, pk):
+class EditPost(UpdateView, Protected):
```
Comment (webknjaz) [inline main/views.py:62]: Do you have tests for this?

File: main/views.py [hunk 18]
```diff
@@ -1,104 +1,136 @@
...
-    post = get_object_or_404(Post, pk=pk)
-    if request.method == 'POST':
+
+    model = Comment
+    fields = ['author', 'text']
+    template_name = 'main/post_edit.html'
+
+    def post(self, request, *args, **kwargs):
+        """Add comment to DB."""
+        my_post = get_object_or_404(Post, pk=kwargs['pk'])
```
Comment (webknjaz) [inline main/views.py]: why `my_post`?

File: main/views.py [hunk 19]
```diff
@@ -1,104 +1,136 @@
...
+class AddCommentToPost(CreateView):
     """If someone wants to create new comment he/she get this view."""
-    post = get_object_or_404(Post, pk=pk)
-    if request.method == 'POST':
+
+    model = Comment
+    fields = ['author', 'text']
+    template_name = 'main/post_edit.html'
+
+    def post(self, request, *args, **kwargs):
```
Comment (webknjaz) [inline main/views.py:104]: you could expose `pk` right here

File: vagga.yaml (hunk 13 above)
Comment (webknjaz) [reply vagga.yaml:28]: okay

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: main/views.py (hunk 14 above)
Comment (DrEdi) [reply main/views.py:16]: I created new class 'Protected' and inherits from this. 
Also I know about standart mixing class for log in, but unfortunately this one was not working for me :\

Comment (webknjaz) [reply main/views.py:16]: what do you mean by inheriting `this`?

Comment (DrEdi) [reply main/views.py:16]: from it*

Comment (webknjaz) [reply main/views.py:16]: but the question was whether `@login_required` **can** be a class decorator

Comment (DrEdi) [reply main/views.py:16]: We can use `@method_decorator` and there put `@login_required`. But IMHO this not very clear way

Comment (webknjaz) [reply main/views.py:16]:
```
@login_required
class BlahBla:
    pass
```

Comment (DrEdi) [reply main/views.py:16]: This is impossible. `@login_required` is only for functions

Comment (webknjaz) [reply main/views.py:16]: That's what I was questioning about

Comment (DrEdi) [reply main/views.py:16]: Nope, we can't :)

File: main/views.py (hunk 18 above)
Comment (DrEdi) [reply main/views.py]: Because 'post' would conflict with method name

File: main/views.py (hunk 17 above)
Comment (DrEdi) [reply main/views.py:62]: Yep, line 75-86 in test_views.py

File: main/views.py (hunk 18 above)
Comment (webknjaz) [reply main/views.py]: really? any proof?

File: main/views.py [hunk 20]
```diff
@@ -24,7 +24,10 @@ class PostList(ListView):
 
     context_object_name = 'posts'
     template_name = 'main/index.html'
-    queryset = Post.objects.filter(published_date__lte=timezone.now())
```
Comment (webknjaz) [inline main/views.py]: doesn't this work?

Comment (DrEdi) [reply main/views.py]: Nope, here we, unfortunately, can't use it.
http://stackoverflow.com/questions/19707237/use-get-queryset-method-or-set-queryset-variable

File: main/views.py (hunk 18 above)
Comment (DrEdi) [reply main/views.py]: I thought :)

**Review by webknjaz — APPROVED** [review]
